### PR TITLE
Pr/pa1

### PR DESCRIPTION
**Pull request by qzmfranklin**
This pull request builds `pa1/pptok.exe` as the preprocessing tokenizer.
You can build it by:

```
cd pa1
make pptok.exe
```

The `pptok.exe` program is largely the same with but differs slightly from the CPPGM `pptoken` program. You can use `compare.py` to compare the results of these two programs.

You can run the CPPGM test suite by

```
./run_tests.sh
```

It uses `compare.py` to compare the results.

As of 01/18/2016, most tests are passing other than some really edge edge cases.

It is only fully tested on my MacBook El Captain so far. It was fully tested on Ubuntu14.04LTS before the `PPTokenizerDFA` commits.

### CODE REVIEW

File: pa1/PPCodeUnitStream.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include "UTF32StreamIfc.h"
+#include "PPCodeUnitStreamIfc.h"
+#include <queue>
+#include <string>
+
+// Stream UTF32 code points into stream of PPCodeUnit that is easy to tokenize.
+//
+//   - Line splicing. i.e., occurrences of "\\\n" are replaced with
+//     PPCodeUnitType::WhitespaceCharacter so that we can revert it in raw strings.
+//
```
Comment (qzmfranklin) [inline pa1/PPCodeUnitStream.h:14]: Comments and digraphs are no longer parsed in the `PPCodeUnitStream` class. Should update the comment to reflect the changes. Might be good to shed light upon why they cannot be treated as `PPCodeUnits`: potential collision with raw string `d-char`.

File: pa1/PPCodeUnitStream.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//     reverted in raw strings.
+class PPCodeUnitStream: public PPCodeUnitStreamIfc {
+public:
+  PPCodeUnitStream(std::shared_ptr<UTF32StreamIfc>);
+
+  virtual bool isEmpty() const override;
+  virtual const std::shared_ptr<PPCodeUnit> getCodeUnit() const override;
+  virtual void toNext() override;
+
+  std::string getErrorMessage() const;
```
Comment (qzmfranklin) [inline pa1/PPCodeUnitStream.h:28]: Should document the following:

`PPCodeUnitStream` nevers throws exceptions, as with the rest of this project. Error reporting and handling is done on the caller site by explicitly checking the error message, as returned by `getErrorMessage()`. If the error message
- is empty, no error occurred in the last call to the `_pushCodeUnits()` function;
- is not empty, some error occurred there.

What a user needs to know:
- It is up to the caller site to print the error message, if any, in any way they want. They may choose to not print it at all.
- The error messages, when printed, are grepable in the source code.

File: pa1/PPCodeUnitStream.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+private:
+  std::shared_ptr<UTF32StreamIfc> _u32stream;
+
+  // Error reporting related stuff.
+  void _setError(const std::string&&);
+  void _clearError();
+  std::string _errorMessage;
+
+  // _pushCodeUnits() parses the input stream and pushes one or more PPCodeUnits
```
Comment (qzmfranklin) [inline pa1/PPCodeUnitStream.h:38]: Maybe I should document the reason for not allowing backward movement of the stream somewhere in the source file:

In summary, no need for it.

Though the instinct are probably that allowing the users of these stream classes, including `UTF32StreamIfc`, `PPCodeUnitStreamIfc`, and `PPTokenizerDFA`, are easier to use with backward movement functions such as `moveOneUnitBackward()` or similar. In reality, the entire preprocessing tokenizer can be constructed without them just fine.

That said, though, it is a long time debate and exploration that I finally found this current interface most suitable. You can move forward by one unit with `toNext()`, get the current unit by `getCodeUnit()`, and check for error messages with `getErrorMessage()`. Note how this same API pattern propagates through `UTF32StreamIfc`, `PPCodeUnitStreamIfc`, and `PPTokenizerDFA` unchanged. As of this moment, I tend to believe that these are the minimal API for getting the work done lazily - that is - done as late as possible through these layers.

File: pa1/PPToken.h [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+public:
+  PPToken(const PPTokenType type): _type(type) {}
+
+  // Interface
+  PPTokenType getType() const { return _type; }
+
+  // Get the corresponding raw text
+  virtual std::string getUTF8String() const = 0;
+
+  // Get human friendly UTF8 string for the given token type
```
Comment (qzmfranklin) [inline pa1/PPToken.h:65]: Should probably document why this function is `static`:

While at first glance it is better to write a non-static pure virtual method `std::string getTypeUTF8String() const = 0` to `PPCodeUnit`, it is indeed much clearer to just list all the preprocessing token types together in a list in the .cpp file. Given that I am using two custom token types, `new-line` and `whitespace-sequence`, this was the least error-prone way of having this functionality when I first developed this structure because I can just copy and paste the list of token types from the C++ specification PDF and know it is guaranteed to be right. Not to mention it saves a lot of coding, compared to having separate methods in each subclass.

File: pa1/PPTokenizerDFA.cpp [hunk 5]
```diff
@@ -0,0 +1,1386 @@
+#include "PPCodePointCheck.h"
+#include "PPCodeUnitCheck.h"
+#include "PPTokenizerDFA.h"
+#include <assert.h>
+
+// TODO: Create a common utility debug logging tool instead of this macro
```
Comment (qzmfranklin) [inline pa1/PPTokenizerDFA.cpp:6]: Yes, I want to do this. But that should be a separate pull request. This pr focuses on preprocessing tokenizer, not utilities as much.

File: pa1/UTF32StreamIfc.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#ifndef UTF32StreamIfc_h
+#define UTF32StreamIfc_h
+
+#include <string>
+
+// Implementation CAVEAT: If the last code point is not the new-line \n, add it.
```
Comment (qzmfranklin) [inline pa1/UTF32StreamIfc.h:6]: Well, this is not exactly a caveat. It is required.

File: pa1/gtest_PPCodePointCheck.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#include "PPCodePointCheck.h"
+#include <gtest/gtest.h>
+
```
Comment (qzmfranklin) [inline pa1/gtest_PPCodePointCheck.cpp:3]: Ah, dumb. Should add ACTUAL tests, shouldn't I?

**Comment by qzmfranklin** [comment]
The CPPGM thinks `<::` is parsed as `<` and `::`. My program parses `<::` as `<:` and `:`. I do not plan to conform to the CPPGM implementation for two reasons:

- The C++ standard does not define the exact behavior for this particular case.
- The lexer has been greedy everywhere else. It makes little sense to not be greedy for only one case.

**Comment by qzmfranklin** [comment]
The CPPGM thinks `<::` is parsed as `<` and `::`. My program parses `<::` as `<:` and `:`. I do not plan to conform to the CPPGM implementation for three reasons:
- The C++ standard does not define the exact behavior for this particular case.
- The lexer has been greedy everywhere else. It makes little sense to not be greedy for only one case.
- This is a rather rare case involving digraphs. The impact of non-conformance to the CPPGM is minimal, if not utterly negligible.

**Comment by qzmfranklin** [comment]
This pr has been around for a while. I am starting to forget things by now.
Passed all tests so far. Consider this pr good now.

**Comment by qzmfranklin** [comment]
Do not delete this branch as some versions of my resumes had pointers to the url of this branch.
